src/hotspot/share/runtime/mutex.cpp

```diff
@@ -522,3 +523,33 @@ void Mutex::set_owner_implementation(Thread *new_owner) {
   }
 }
 #endif // ASSERT
+
+
+RecursiveMutex::RecursiveMutex() : _sem(1), _owner(nullptr), _recursions(0) {}
+
+void RecursiveMutex::lock(Thread* current) {
+  assert(current == Thread::current(), "must be current thread");
+  if (current == _owner) {
+    _recursions++;
+  } else {
+    // can be called by jvmti by VMThread.
+    if (current->is_Java_thread()) {
+      _sem.wait_with_safepoint_check(JavaThread::cast(current));
```

Comment:
Why not use PlatformMutex + OSThreadWaitState instead of a semaphore?

Reply:
Semaphore seemed simpler (?)

Reply:
OK.  It's a good thing HotSpot doesn't need to worry about priority-inheritance for mutexes.

Reply:
Semaphore seems simpler/cleaner and ready to use.

Reply:
It's such a rare race and unusual condition that we could execute more Java code, that I started out with just a shared bit.  Then David suggested a semaphore that obeys the safepoint protocol, which seems a lot better.  I've literally skimmed over OSThreadWaitState.  It looks like Semaphore::wait_with_a_safepoint_check() uses it.  I still don't know why it exists:

```
// Note: the ThreadState is legacy code and is not correctly implemented.
// Uses of ThreadState need to be replaced by the state in the JavaThread.

enum ThreadState {
```

Does a PlatformMutex handle priority-inheritance? It still feels like it would be overkill for this usage.  I hope this RecursiveLocker does not become more widely used, where we would have to replace the simple implementation with something more difficult.  We should discourage further use when possible.

Reply:
Thanks for your last comment because I was worried there's a lot of other code I should know about.